src/Hadrian/Builder.hs

```diff
@@ -89,25 +94,37 @@ build = buildWith [] []
 buildWithResources :: (Builder b, ShakeValue c) => [(Resource, Int)] -> Target c b -> Args c b -> Action ()
 buildWithResources rs = buildWith rs []
 
+askWithResources :: (Builder b, ShakeValue c) => [(Resource, Int)] -> Target c b -> Args c b -> Action String
+askWithResources rs = askWith rs []
+
 -- | Like 'build' but passes given options to Shake's 'cmd'.
 buildWithCmdOptions :: (Builder b, ShakeValue c) => [CmdOption] -> Target c b -> Args c b -> Action ()
 buildWithCmdOptions = buildWith []
 
-buildWith :: (Builder b, ShakeValue c) => [(Resource, Int)] -> [CmdOption] -> Target c b -> Args c b -> Action ()
-buildWith rs opts target args = do
+doWith :: (Builder b, ShakeValue c)
+       => (b -> BuildInfo -> Action a)
+       -> (Target c b -> Action ())
+       -> [(Resource, Int)] -> [CmdOption] -> Target c b -> Args c b -> Action a
+doWith f info rs opts target args = do
     needBuilder (builder target)
     argList <- interpret target args
     trackArgsHash target -- Rerun the rule if the hash of argList has changed.
-    putInfo target
+    info target
     verbose <- interpret target verboseCommand
     let quietlyUnlessVerbose = if verbose then withVerbosity Loud else quietly
-    quietlyUnlessVerbose $ runBuilderWith (builder target) $
+    quietlyUnlessVerbose $ f (builder target) $
         BuildInfo { buildArgs      = argList
                   , buildInputs    = inputs target
                   , buildOutputs   = outputs target
                   , buildOptions   = opts
                   , buildResources = rs }
 
+buildWith :: (Builder b, ShakeValue c) => [(Resource, Int)] -> [CmdOption] -> Target c b -> Args c b -> Action ()
+buildWith = doWith runBuilderWith runInfo
+
+askWith :: (Builder b, ShakeValue c) => [(Resource, Int)] -> [CmdOption] -> Target c b -> Args c b -> Action String
+askWith = doWith askBuilderWith askInfo
+
 -- | Print out information about the command being executed.
 putInfo :: Show b => Target c b -> Action ()
```

Comment:
Is this dead code? Looks like we should have simply renamed `putInfo` to `runInfo`.

Reply:
`putInfo` indeed isn't used anywhere. It's used in a single place on `master`. I'm not sure I understand what you want me to change here though?

Reply:
Well, if `putInfo` is not used any more, let's just delete it.

Reply:
Done in an upcoming commit.